### PR TITLE
Add complex128 support

### PR DESCRIPTION
**Pull request by hunkim**
Added the complex128 support, based on the guidance from @girving and others in PR #2244. All tests (in CPU) including backwards_compatibility_test are passed! I couldn't test the GPU version.

The complex128 support is not added for _FFT_ ops with input: complex64 and output: complex64.

### CODE REVIEW

File: tensorflow/core/ops/math_ops.cc [hunk 1]
```diff
@@ -291,19 +291,19 @@ Returns element-wise smallest integer in not less than x.
 
 #define BINARY_MORE()                              \
   Input("x: T").Input("y: T").Output("z: T").Attr( \
-      "T: {half, float, double, uint8, int8, int16, int32, int64, complex64}")
+      "T: {half, float, double, uint8, int8, int16, int32, int64, complex64, complex128}")
```
Comment (girving) [inline tensorflow/core/ops/math_ops.cc]: 80 columns

File: tensorflow/core/ops/math_ops.cc [hunk 2]
```diff
@@ -291,19 +291,19 @@ Returns element-wise smallest integer in not less than x.
...
+      "T: {half, float, double, int32, int64, complex64, complex128}")
 
 // TODO(mrry): Restore `SetIsCommutative()` for non-string types.
 REGISTER_OP("Add")
     .Input("x: T")
     .Input("y: T")
     .Output("z: T")
     .Attr(
-        "T: {half, float, double, uint8, int8, int16, int32, int64, complex64, "
+        "T: {half, float, double, uint8, int8, int16, int32, int64, complex64, complex128, "
```
Comment (girving) [inline tensorflow/core/ops/math_ops.cc]: 80 columns

File: tensorflow/core/ops/math_ops.cc [hunk 3]
```diff
@@ -1161,7 +1161,7 @@ tf.complex(real, imag) ==> [[2.25 + 4.75j], [3.25 + 5.75j]]
 ```
 )doc");
 
-REGISTER_OP("Real").Input("input: complex64").Output("output: float").Doc(R"doc(
+REGISTER_OP("Real").Input("input: complex64, complex128").Output("output: float").Doc(R"doc(
```
Comment (girving) [inline tensorflow/core/ops/math_ops.cc]: 80 columns

File: tensorflow/core/ops/math_ops.cc [hunk 4]
```diff
@@ -1177,7 +1177,7 @@ tf.real(input) ==> [-2.25, 3.25]
 ```
 )doc");
 
-REGISTER_OP("Imag").Input("input: complex64").Output("output: float").Doc(R"doc(
+REGISTER_OP("Imag").Input("input: complex64, complex128").Output("output: float").Doc(R"doc(
```
Comment (girving) [inline tensorflow/core/ops/math_ops.cc]: 80 columns

**Comment by girving** [comment]
Please squash into one commit.  It looks like a lot of the intermediate commits don't even compile, much less work, which makes them harder to review.

**Comment by hunkim** [comment]
@girving Thanks! All commits are squashed, and lines more than 80 columns are fixed.
Four lines (81/82 columns) in tensorflow/core/kernels/cwise_ops_common.h are not fixed for better readability.

All my local 457 tests pass.

**Comment by girving** [comment]
Please wrap the 81/82 column lines.  It's a hard limit, except for Python import statements.

File: tensorflow/core/kernels/cwise_op_gpu_select.cu.cc [hunk 5]
```diff
@@ -77,6 +77,7 @@ SELECT_FUNCTOR(double);
 SELECT_FUNCTOR(int32);
 SELECT_FUNCTOR(int64);
 SELECT_FUNCTOR(complex64);
+SELECT_FUNCTOR(complex6128);
```
Comment (girving) [inline tensorflow/core/kernels/cwise_op_gpu_select.cu.cc]: This seems unlikely to work.

**Comment by girving** [comment]
Looks good, thanks!  I'll fire off the tests once you fix the `complex6128` bit and the > 80 column lines.

File: tensorflow/core/kernels/cwise_op_gpu_select.cu.cc (hunk 5 above)
Comment (hunkim) [inline tensorflow/core/kernels/cwise_op_gpu_select.cu.cc]: Oops! Fixed.

**Comment by hunkim** [comment]
Done!

**Comment by girving** [comment]
Woot!  Jenkins, test this please.

**Comment by girving** [comment]
Jenkins, test this please.

**Comment by hunkim** [comment]
Sorry that I could not run GPU tests locally. But I guess now it should pass GPU Tests.

**Comment by zffchen78** [comment]
Is there any test excercising the new type complex128's operations?

**Comment by hunkim** [comment]
@zffchen78, I don't think we have tests for complex128 for now, but we will have some very soon.

**Comment by zffchen78** [comment]
The code base is fast changing. The only way to ensure it works for everyone is to maintain a high standard of the code quality. One important aspect is to make sure testing is done well. In your case, can you add proper test cases covering the operation this PR touches? It should not be too complicated. Most relevant tests should be in python/kernel_tests/cwise_ops_test.py. You can grep test cases for complex64, most of which is just matter of extending the type list w/ the new type.

**Comment by hunkim** [comment]
@zffchen78, thanks! I agree and I'll work on it.

**Comment by hunkim** [comment]
@girving Do you have more comments on this PR? I'd be happy to address your comments.

**Comment by girving** [comment]
@hunkim: Apologies, I should have clarified.  I agree with @zffchen78 that you should add tests, and was waiting on those.  It should be as simple as searching for `complex64` in `cwise_ops_test.py` and making sure `complex128` does the same thing.

**Comment by hunkim** [comment]
@girving Sounds good. I'm on it and will update soon.

**Comment by hunkim** [comment]
Woot! We have a new PR #2417.

**Pull request closed without merging**